Requirement failed warning while running BQSR @DecadentRead

When running BQSR with the following command,

adam-submit transform /small.adam /small_bqsr.adam -recalibrate_base_qualities -known_snps /db/dbsnp_137.hg19.excluding_sites_after_129.vcf --master yarn-client --num-executors 85

I still got the following 'requirement failed Error' in the DecadentRead and there is no output for the BQSR.

15/04/09 13:58:17 WARN TaskSetManager: Lost task 0.0 in stage 1.0 (TID 18, node-119): java.lang.IllegalArgumentException: Error "requirement failed: sequence and qualityScores must be same length" while constructing DecadentRead from Read({"contig": {"contigName": "1", "contigLength": 249250621, "contigMD5": null, "referenceURL": null, "assembly": null, "species": null}, "start": 26472783, "oldPosition": null, "end": 26472858, "mapq": 60, "readName": "simread:1:26472783:false", "sequence": "GTATAAGAGCAGCCTTATTCCTATTTATAATCAGGGTGAAACACCTGTGCCAATGCCAAGACAGGGGTGCCAAGA", "qual": "*", "cigar": "75M", "oldCigar": null, "basesTrimmedFromStart": 0, "basesTrimmedFromEnd": 0, "readPaired": false, "properPair": false, "readMapped": true, "mateMapped": false, "firstOfPair": false, "secondOfPair": false, "failedVendorQualityChecks": false, "duplicateRead": false, "readNegativeStrand": true, "mateNegativeStrand": false, "primaryAlignment": true, "secondaryAlignment": false, "supplementaryAlignment": false, "mismatchingPositions": null, "origQual": null, "attributes": "XS:i:0\tAS:i:75\tNM:i:0", "recordGroupName": null, "recordGroupSequencingCenter": null, "recordGroupDescription": null, "recordGroupRunDateEpoch": null, "recordGroupFlowOrder": null, "recordGroupKeySequence": null, "recordGroupLibrary": null, "recordGroupPredictedMedianInsertSize": null, "recordGroupPlatform": null, "recordGroupPlatformUnit": null, "recordGroupSample": null, "mateAlignmentStart": null, "mateAlignmentEnd": null, "mateContig": null})
	at org.bdgenomics.adam.rich.DecadentRead$.apply(DecadentRead.scala:40)
	at org.bdgenomics.adam.rich.DecadentRead$.apply(DecadentRead.scala:32)
	at org.bdgenomics.adam.rich.DecadentRead$$anonfun$cloy$1.apply(DecadentRead.scala:50)
	at org.bdgenomics.adam.rich.DecadentRead$$anonfun$cloy$1.apply(DecadentRead.scala:50)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:328)
	at scala.collection.Iterator$$anon$14.hasNext(Iterator.scala:389)
	at scala.collection.Iterator$$anon$13.hasNext(Iterator.scala:371)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:327)
	at scala.collection.Iterator$class.foreach(Iterator.scala:727)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1157)
	at scala.collection.TraversableOnce$class.foldLeft(TraversableOnce.scala:144)
	at scala.collection.AbstractIterator.foldLeft(Iterator.scala:1157)
	at scala.collection.TraversableOnce$class.aggregate(TraversableOnce.scala:201)
	at scala.collection.AbstractIterator.aggregate(Iterator.scala:1157)
	at org.apache.spark.rdd.RDD$$anonfun$22.apply(RDD.scala:901)
	at org.apache.spark.rdd.RDD$$anonfun$22.apply(RDD.scala:901)
	at org.apache.spark.SparkContext$$anonfun$29.apply(SparkContext.scala:1355)
	at org.apache.spark.SparkContext$$anonfun$29.apply(SparkContext.scala:1355)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:61)
	at org.apache.spark.scheduler.Task.run(Task.scala:56)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:196)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalArgumentException: requirement failed: sequence and qualityScores must be same length
	at scala.Predef$.require(Predef.scala:233)
	at org.bdgenomics.adam.rich.DecadentRead.<init>(DecadentRead.scala:61)
	at org.bdgenomics.adam.rich.DecadentRead$.apply(DecadentRead.scala:36)
	... 23 more

--- a/adam-core/src/main/scala/org/bdgenomics/adam/converters/AlignmentRecordConverter.scala
+++ b/adam-core/src/main/scala/org/bdgenomics/adam/converters/AlignmentRecordConverter.scala
@@ -79,7 +79,10 @@ class AlignmentRecordConverter extends Serializable {
     // set canonically necessary fields
     builder.setReadName(adamRecord.getReadName.toString)
     builder.setReadString(adamRecord.getSequence)
-    builder.setBaseQualityString(adamRecord.getQual)
+    adamRecord.getQual match {
+      case null      => builder.setBaseQualityString("*")
+      case s: String => builder.setBaseQualityString(s)
+    }
 
     // set read group flags
     Option(adamRecord.getRecordGroupName)

--- a/adam-core/src/main/scala/org/bdgenomics/adam/converters/SAMRecordConverter.scala
+++ b/adam-core/src/main/scala/org/bdgenomics/adam/converters/SAMRecordConverter.scala
@@ -65,8 +65,14 @@ class SAMRecordConverter extends Serializable with Logging {
         .setCigar(cigar)
         .setBasesTrimmedFromStart(startTrim)
         .setBasesTrimmedFromEnd(endTrim)
-        .setQual(samRecord.getBaseQualityString)
         .setOrigQual(SAMUtils.phredToFastq(samRecord.getOriginalBaseQualities))
+
+      // if the quality string is "*", then we null it in the record
+      // or, in other words, we only set the quality string if it is not "*"
+      val qual = samRecord.getBaseQualityString
+      if (qual != "*") {
+        builder.setQual(qual)
+      }
 
       // Only set the reference information if the read is aligned, matching the mate reference
       // This prevents looking up a -1 in the sequence dictionary

--- a/adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/recalibration/BaseQualityRecalibration.scala
+++ b/adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/recalibration/BaseQualityRecalibration.scala
@@ -55,6 +55,7 @@ class BaseQualityRecalibration(
   val dataset: RDD[(CovariateKey, Residue)] = {
     def shouldIncludeRead(read: DecadentRead) =
       read.isCanonicalRecord &&
+        read.record.record.getQual != null &&
         read.alignmentQuality.exists(_ > QualityScore.zero) &&
         read.passedQualityChecks
 

--- a/adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/recalibration/Recalibrator.scala
+++ b/adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/recalibration/Recalibrator.scala
@@ -30,10 +30,14 @@ class Recalibrator(val table: RecalibrationTable, val minAcceptableQuality: Qual
 
   def apply(read: DecadentRead): AlignmentRecord = RecalibrateRead.time {
     val record: AlignmentRecord = read.record
-    AlignmentRecord.newBuilder(record).
-      setQual(QualityScore.toString(computeQual(read))).
-      setOrigQual(record.getQual).
-      build()
+    if (record.getQual != null) {
+      AlignmentRecord.newBuilder(record)
+        .setQual(QualityScore.toString(computeQual(read)))
+        .setOrigQual(record.getQual)
+        .build()
+    } else {
+      record
+    }
   }
 
   def computeQual(read: DecadentRead): Seq[QualityScore] = ComputeQualityScore.time {

--- a/adam-core/src/main/scala/org/bdgenomics/adam/rich/DecadentRead.scala
+++ b/adam-core/src/main/scala/org/bdgenomics/adam/rich/DecadentRead.scala
@@ -58,7 +58,8 @@ class DecadentRead(val record: RichAlignmentRecord) extends Logging {
   require(!record.getPrimaryAlignment || record.getReadMapped, "Unaligned read can't be a primary alignment")
 
   // Should have quality scores for all residues
-  require(record.getSequence.length == record.qualityScores.length, "sequence and qualityScores must be same length")
+  require(record.getQual == null ||
+    record.getSequence.length == record.qualityScores.length, "sequence and qualityScores must be same length")
 
   // MapQ should be valid
   require(record.getMapq == null || (record.getMapq >= 0 && record.getMapq <= 93), "MapQ must be in [0, 255]")

--- a/adam-core/src/test/scala/org/bdgenomics/adam/converters/AlignmentRecordConverterSuite.scala
+++ b/adam-core/src/test/scala/org/bdgenomics/adam/converters/AlignmentRecordConverterSuite.scala
@@ -32,9 +32,9 @@ class AlignmentRecordConverterSuite extends FunSuite {
   // allocate converters
   val adamRecordConverter = new AlignmentRecordConverter
 
-  def make_read(start: Long, cigar: String, mdtag: String, length: Int, id: Int = 0): AlignmentRecord = {
+  def make_read(start: Long, cigar: String, mdtag: String, length: Int, id: Int = 0, nullQuality: Boolean = false): AlignmentRecord = {
     val sequence: String = "A" * length
-    AlignmentRecord.newBuilder()
+    val builder = AlignmentRecord.newBuilder()
       .setReadName("read" + id.toString)
       .setStart(start)
       .setReadMapped(true)
@@ -42,11 +42,15 @@ class AlignmentRecordConverterSuite extends FunSuite {
       .setSequence(sequence)
       .setReadNegativeStrand(false)
       .setMapq(60)
-      .setQual(sequence) // no typo, we just don't care
       .setMismatchingPositions(mdtag)
       .setOldPosition(12)
       .setOldCigar("2^AAA3")
-      .build()
+
+    if (!nullQuality) {
+      builder.setQual(sequence) // no typo, we just don't care
+    }
+
+    builder.build()
   }
 
   test("testing the fields in a converted ADAM Read") {
@@ -88,6 +92,54 @@ class AlignmentRecordConverterSuite extends FunSuite {
     assert(toSAM.getReadNegativeStrandFlag === false)
     assert(toSAM.getMappingQuality === 60)
     assert(toSAM.getBaseQualityString === sequence)
+    assert(toSAM.getAttribute("MD") === "2^AAA2")
+    assert(toSAM.getIntegerAttribute("OP") === 13)
+    assert(toSAM.getStringAttribute("OC") === "2^AAA3")
+    //make sure that we didn't set the SM attribute.
+    //issue #452 https://github.com/bigdatagenomics/adam/issues/452
+    assert(toSAM.getAttribute("SM") === null)
+    assert(toSAM.getHeader().getReadGroup("record_group").getSample() === "sample")
+  }
+
+  test("converting a read with null quality is OK") {
+    val adamRead = make_read(3L, "2M3D2M", "2^AAA2", 4, nullQuality = true)
+
+    // add reference details
+    adamRead.setRecordGroupName("record_group")
+    adamRead.setRecordGroupSample("sample")
+    adamRead.setContig(Contig.newBuilder()
+      .setContigName("referencetest")
+      .build())
+    adamRead.setMateContig(Contig.newBuilder()
+      .setContigName("matereferencetest")
+      .setContigLength(6L)
+      .setReferenceURL("test://chrom1")
+      .build())
+    adamRead.setMateAlignmentStart(6L)
+
+    // make sequence dictionary
+    val seqRecForDict = SequenceRecord("referencetest", 5, "test://chrom1")
+    val dict = SequenceDictionary(seqRecForDict)
+
+    //make read group dictionary
+    val readGroup = new RecordGroup(adamRead.getRecordGroupSample(), adamRead.getRecordGroupName())
+    val readGroups = new RecordGroupDictionary(Seq(readGroup))
+
+    // convert read
+    val toSAM = adamRecordConverter.convert(adamRead,
+      SAMFileHeaderWritable(adamRecordConverter.createSAMHeader(dict,
+        readGroups)))
+
+    // validate conversion
+    val sequence = "A" * 4
+    assert(toSAM.getReadName === ("read" + 0.toString))
+    assert(toSAM.getAlignmentStart === 4)
+    assert(toSAM.getReadUnmappedFlag === false)
+    assert(toSAM.getCigarString === "2M3D2M")
+    assert(toSAM.getReadString === sequence)
+    assert(toSAM.getReadNegativeStrandFlag === false)
+    assert(toSAM.getMappingQuality === 60)
+    assert(toSAM.getBaseQualityString === "*")
     assert(toSAM.getAttribute("MD") === "2^AAA2")
     assert(toSAM.getIntegerAttribute("OP") === 13)
     assert(toSAM.getStringAttribute("OC") === "2^AAA3")

--- a/adam-core/src/test/scala/org/bdgenomics/adam/converters/SAMRecordConverterSuite.scala
+++ b/adam-core/src/test/scala/org/bdgenomics/adam/converters/SAMRecordConverterSuite.scala
@@ -17,12 +17,11 @@
  */
 package org.bdgenomics.adam.converters
 
+import htsjdk.samtools._
+import java.io.File
 import org.scalatest.FunSuite
 import org.bdgenomics.adam.models.{ SequenceRecord, Attribute, RecordGroupDictionary, SequenceDictionary }
 import org.bdgenomics.adam.rdd.ADAMContext._
-import java.io.File
-import htsjdk.samtools._
-
 import scala.collection.JavaConversions._
 
 class SAMRecordConverterSuite extends FunSuite {
@@ -33,30 +32,29 @@ class SAMRecordConverterSuite extends FunSuite {
     val testFileString = getClass.getClassLoader.getResource("reads12.sam").getFile
     val testFile = new File(testFileString)
 
-    //Iterator of SamReads in the file that each have a samRecord for conversion
+    // Iterator of SamReads in the file that each have a samRecord for conversion
     val testIterator = SamReaderFactory.makeDefault().open(testFile)
     val testSAMRecord = testIterator.iterator().next()
 
-    //Creating the two SequenceRecords in file for SequenceDictionary
+    // Creating the two SequenceRecords in file for SequenceDictionary
     val testSequenceRecord1 = new SequenceRecord("1", 249250621L)
     val testSequenceRecord2 = new SequenceRecord("2", 243199373L)
 
-    //SequenceDictionary to be used as parameter during conversion
+    // SequenceDictionary to be used as parameter during conversion
     val testSequenceDict = SequenceDictionary(testSequenceRecord1, testSequenceRecord2)
 
-    //RecordGroupDictionary to be used as a parameter during conversion
+    // RecordGroupDictionary to be used as a parameter during conversion
     val testRecordGroupDict = new RecordGroupDictionary(Seq())
 
-    //Convert samRecord to alignmentRecord
+    // Convert samRecord to alignmentRecord
     val testAlignmentRecord = testRecordConverter.convert(testSAMRecord, testSequenceDict, testRecordGroupDict)
 
-    //Validating Conversion
+    // Validating Conversion
     assert(testAlignmentRecord.getCigar === testSAMRecord.getCigarString)
     assert(testAlignmentRecord.getDuplicateRead === testSAMRecord.getDuplicateReadFlag)
     assert(testAlignmentRecord.getEnd.toInt === testSAMRecord.getAlignmentEnd)
     assert(testAlignmentRecord.getMapq.toInt === testSAMRecord.getMappingQuality)
     assert(testAlignmentRecord.getStart.toInt === (testSAMRecord.getAlignmentStart - 1))
-    assert(testAlignmentRecord.getQual === testSAMRecord.getBaseQualityString)
     assert(!testAlignmentRecord.getFirstOfPair)
     assert(testAlignmentRecord.getFailedVendorQualityChecks === testSAMRecord.getReadFailsVendorQualityCheckFlag)
     assert(!testAlignmentRecord.getPrimaryAlignment === testSAMRecord.getNotPrimaryAlignmentFlag)
@@ -66,47 +64,78 @@ class SAMRecordConverterSuite extends FunSuite {
     assert(!testAlignmentRecord.getReadPaired)
     assert(!testAlignmentRecord.getSecondOfPair)
     assert(testAlignmentRecord.getSupplementaryAlignment === testSAMRecord.getSupplementaryAlignmentFlag)
-
   }
 
   test("testing the fields in an alignmentRecord obtained from an unmapped samRecord conversion") {
+
+    val testRecordConverter = new SAMRecordConverter
+    val testFileString = getClass.getClassLoader.getResource("reads12.sam").getFile
+    val testFile = new File(testFileString)
+
+    // Iterator of SamReads in the file that each have a samRecord for conversion
+    val testIterator = SamReaderFactory.makeDefault().open(testFile)
+    val testSAMRecord = testIterator.iterator().next()
+
+    // Creating the two SequenceRecords in file for SequenceDictionary
+    val testSequenceRecord1 = new SequenceRecord("1", 249250621L)
+    val testSequenceRecord2 = new SequenceRecord("2", 243199373L)
+
+    // SequenceDictionary to be used as parameter during conversion
+    val testSequenceDict = SequenceDictionary(testSequenceRecord1, testSequenceRecord2)
+
+    // RecordGroupDictionary to be used as a parameter during conversion
+    val testRecordGroupDict = new RecordGroupDictionary(Seq())
+
+    // Convert samRecord to alignmentRecord
+    val testAlignmentRecord = testRecordConverter.convert(testSAMRecord, testSequenceDict, testRecordGroupDict)
+
+    // Validating Conversion
+    assert(testAlignmentRecord.getCigar === testSAMRecord.getCigarString)
+    assert(testAlignmentRecord.getDuplicateRead === testSAMRecord.getDuplicateReadFlag)
+    assert(testAlignmentRecord.getEnd.toInt === testSAMRecord.getAlignmentEnd)
+    assert(testAlignmentRecord.getMapq.toInt === testSAMRecord.getMappingQuality)
+    assert(testAlignmentRecord.getStart.toInt === (testSAMRecord.getAlignmentStart - 1))
+    assert(!testAlignmentRecord.getFirstOfPair)
+    assert(testAlignmentRecord.getFailedVendorQualityChecks === testSAMRecord.getReadFailsVendorQualityCheckFlag)
+    assert(!testAlignmentRecord.getPrimaryAlignment === testSAMRecord.getNotPrimaryAlignmentFlag)
+    assert(!testAlignmentRecord.getReadMapped === testSAMRecord.getReadUnmappedFlag)
+    assert(testAlignmentRecord.getReadName === testSAMRecord.getReadName)
+    assert(testAlignmentRecord.getReadNegativeStrand === testSAMRecord.getReadNegativeStrandFlag)
+    assert(!testAlignmentRecord.getReadPaired)
+    assert(!testAlignmentRecord.getSecondOfPair)
+    assert(testAlignmentRecord.getSupplementaryAlignment === testSAMRecord.getSupplementaryAlignmentFlag)
+  }
+
+  test("'*' quality gets nulled out") {
 
     val newRecordConverter = new SAMRecordConverter
     val newTestFile = new File(getClass.getClassLoader.getResource("unmapped.sam").getFile)
     val newSAMReader = SamReaderFactory.makeDefault().open(newTestFile)
 
-    //Obtain SAMRecord
+    // Obtain SAMRecord
     val newSAMRecordIter = {
       val samIter = asScalaIterator(newSAMReader.iterator())
       samIter.toIterable.dropWhile(!_.getReadUnmappedFlag())
     }
     val newSAMRecord = newSAMRecordIter.toIterator.next()
 
-    //SequenceRecord to be put into SequenceDictionary
+    // null out quality
+    newSAMRecord.setBaseQualityString("*")
+
+    // SequenceRecord to be put into SequenceDictionary
     val newSequenceRecord = new SequenceRecord("1", 249250621L)
 
-    //SequenceDictionary for conversion method
+    // SequenceDictionary for conversion method
     val newSequenceDictionary = SequenceDictionary(newSequenceRecord)
 
-    //No RecordGroup present in file so I'll just make the RecordGroupDictionary Empty
+    // No RecordGroup present in file so I'll just make the RecordGroupDictionary Empty
     val newTestRecordGroupDictionary = new RecordGroupDictionary(Seq())
 
-    //Conversion
+    // Conversion
     val newAlignmentRecord = newRecordConverter.convert(newSAMRecord, newSequenceDictionary, newTestRecordGroupDictionary)
 
-    //Validating Conversion
-    assert(newAlignmentRecord.getCigar === newSAMRecord.getCigarString)
-    assert(newAlignmentRecord.getDuplicateRead === newSAMRecord.getDuplicateReadFlag)
-    assert(newAlignmentRecord.getEnd === null)
-    assert(newAlignmentRecord.getMapq === null)
-    assert(newAlignmentRecord.getStart === null)
-    assert(newAlignmentRecord.getQual === newSAMRecord.getBaseQualityString)
-    assert(!newAlignmentRecord.getFirstOfPair)
-    assert(newAlignmentRecord.getFailedVendorQualityChecks === newSAMRecord.getReadFailsVendorQualityCheckFlag)
-    assert(!newAlignmentRecord.getReadMapped === newSAMRecord.getReadUnmappedFlag)
-    assert(newAlignmentRecord.getReadName === newSAMRecord.getReadName)
-    assert(newAlignmentRecord.getReadNegativeStrand === newSAMRecord.getReadNegativeStrandFlag)
-
+    // Validating Conversion
+    assert(newAlignmentRecord.getQual === null)
   }
 
 }

--- a/adam-core/src/test/scala/org/bdgenomics/adam/rich/DecadentReadSuite.scala
+++ b/adam-core/src/test/scala/org/bdgenomics/adam/rich/DecadentReadSuite.scala
@@ -68,4 +68,21 @@ class DecadentReadSuite extends FunSuite {
     assert(residueSeq.head.referencePosition === ReferencePosition("chr1", 1000))
   }
 
+  test("build a decadent read from a read with null qual") {
+    val contig = Contig.newBuilder
+      .setContigName("chr1")
+      .build
+
+    val hardClippedRead = RichAlignmentRecord(AlignmentRecord
+      .newBuilder()
+      .setReadMapped(true)
+      .setStart(1000)
+      .setContig(contig)
+      .setMismatchingPositions("10")
+      .setSequence("AACCTTGGC")
+      .setCigar("9M1H").build())
+
+    val record = DecadentRead(hardClippedRead)
+    assert(record.residues.size === 9)
+  }
 }

